src/vulkan/Context.zig

```diff
@@ -72,13 +74,124 @@ pub fn init(
+    const device_properties = try self.instance.enumerateDeviceExtensionPropertiesAlloc(pdev, null, allocator);
+    defer allocator.free(device_properties);
+
+    for (instance_extensions) |extension| {
+        for (device_properties) |property| {
+            if (std.mem.eql(u8, property.extension_name, extension))
+                break;
+        } else {
+            return error.ExtensionNotPresent;
+        }
+    }
+    return device_properties.len;
+}
+
+fn checkSurfaceSupport(self: *Self, pdev: vk.PhysicalDevice, surface: vk.SurfaceKHR, allocator: Allocator) !u32 {
+    _ = allocator;
+    var format_count: u32 = undefined;
+    _ = try self.instance.getPhysicalDeviceSurfaceFormatsKHR(pdev, surface, &format_count, null);
+
+    var present_mode_count: u32 = undefined;
+    _ = try self.instance.getPhysicalDeviceSurfacePresentModesKHR(pdev, surface, &present_mode_count, null);
+
+    if (format_count > 0 and present_mode_count > 0)
+        return format_count + present_mode_count;
+    return error.FeatureNotPresent;
+}
+
+fn checkDeviceQueueSupport(self: *Self, pdev: vk.PhysicalDevice, surface: vk.SurfaceKHR, allocator: Allocator) !u64 {
+    const families = try self.instance.getPhysicalDeviceQueueFamilyPropertiesAlloc(pdev, allocator);
+    defer allocator.free(families);
+
+    var graphics_family: ?u32 = null;
+    var present_family: ?u32 = null;
+
+    for (families, 0..) |properties, i| {
+        const family: u32 = @intCast(i);
+
+        if (graphics_family == null and properties.queue_flags.graphics_bit) {
+            graphics_family = family;
+        }
+
+        if (present_family == null and (try self.instance.getPhysicalDeviceSurfaceSupportKHR(pdev, family, surface)) == vk.TRUE) {
+            present_family = family;
+        }
+    }
+
+    if (graphics_family != null and present_family != null) {
+        return families.len;
+    }
+
+    return error.FeatureNotPresent;
+}
+
+fn allocDeviceQueues(self: *Self, pdev: vk.PhysicalDevice, surface: vk.SurfaceKHR, allocator: Allocator) !void {
+    const families = try self.instance.getPhysicalDeviceQueueFamilyPropertiesAlloc(pdev, allocator);
+    defer allocator.free(families);
+
+    var graphics_family: ?u32 = null;
+    var present_family: ?u32 = null;
+
+    for (families, 0..) |properties, i| {
+        const family: u32 = @intCast(i);
+
+        if (graphics_family == null and properties.queue_flags.graphics_bit) {
+            graphics_family = family;
+        }
+
+        if (present_family == null and (try self.instance.getPhysicalDeviceSurfaceSupportKHR(pdev, family, surface)) == vk.TRUE) {
+            present_family = family;
+        }
+    }
+
+    if (graphics_family != null and present_family != null) {
+        self.queue_families = QueueFamilies{ .graphics_queue = graphics_family.?, .present_queue = present_family.? };
+    }
+
+    // sanity check
+    return error.FeatureNotPresent;
```

Comment:
this unconditionally fails the function, intended to `return;` ?

Reply:
yes, resolved while reviewing

Reply:
no return needed, since !void function